tool-openssl/x509.cc

```diff
@@ -3,72 +3,251 @@
 
 #include <openssl/x509.h>
 #include <openssl/pem.h>
+#include <openssl/evp.h>
+#include <openssl/rsa.h>
+#include <openssl/bn.h>
+#include <chrono>
+#include <cstdio>
+#include <ctime>
 #include "internal.h"
 
 static const argument_t kArguments[] = {
-  { "-in", kRequiredArgument, "Input file" },
-  { "-out", kRequiredArgument, "Output file" },
-  { "", kOptionalArgument, "" }
+    { "-in", kRequiredArgument, "Input file" },
+    { "-out", kOptionalArgument, "Output file" },
+    { "-noout", kBooleanArgument, "No output" },
+    { "-modulus", kBooleanArgument, "Modulus" },
+    { "-signkey", kOptionalArgument, "Sign key" },
+    { "-days", kOptionalArgument, "Days" },
+    { "-dates", kBooleanArgument, "Dates" },
+    { "-req", kBooleanArgument, "Req" },
+    { "-checkend", kOptionalArgument, "Checkend" },
+    { "", kOptionalArgument, "" }
 };
 
+bool WriteSignedCertificate(X509 *x509, const std::string &out_path) {
+  ScopedFILE out_file(fopen(out_path.c_str(), "wb"));
+  if (!out_file) {
+    fprintf(stderr, "Error: unable to open output file '%s'\n", out_path.c_str());
+    return false;
+  }
+  if (!PEM_write_X509(out_file.get(), x509)) {
+    fprintf(stderr, "Error: error writing certificate to '%s'\n", out_path.c_str());
+    ERR_print_errors_fp(stderr);
+    return false;
+  }
+  return true;
+}
+
+bool LoadAndSignCertificate(X509 *x509, const std::string &signkey_path) {
+  ScopedFILE signkey_file(fopen(signkey_path.c_str(), "rb"));
+  if (!signkey_file) {
+    fprintf(stderr, "Error: unable to load private key from '%s'\n", signkey_path.c_str());
+    return false;
+  }
+  EVP_PKEY *pkey = PEM_read_PrivateKey(signkey_file.get(), nullptr, nullptr, nullptr);
+  if (!pkey) {
+    fprintf(stderr, "Error: error reading private key from '%s'\n", signkey_path.c_str());
+    ERR_print_errors_fp(stderr);
+    return false;
+  }
+  bssl::UniquePtr<EVP_PKEY> pkey_guard(pkey);
+  if (!X509_sign(x509, pkey, EVP_sha256())) {
```

Comment:
Just double checking that OpenSSL's default is SHA256 right? This isn't part of this project phase, but it might be customizable with the `-digest` option later on. 
Could we leave a TODO comment above here to help us keep track?